File: day01/main.ex
defmodule Day01 do
  def solve_part1(input) do
    input
    |> Enum.map(
      &case &1 do
        "(" -> 1
        ")" -> -1
      end
    )
    |> Enum.sum()
  end

  def solve_part2(contents) do
    contents
    |> Enum.with_index()
    |> Enum.reduce({0, -1}, fn {char, index}, {floor, basement} ->
      case char do
        "(" ->
          {floor + 1, basement}

        ")" ->
          {floor - 1,
           if basement === -1 && floor === 0 do
             index + 1
           else
             basement
           end}
      end
    end)
    |> elem(1)
  end
end

contents =
  File.read!("day01/input.txt")
  |> String.trim()
  |> String.graphemes()

contents
|> Day01.solve_part1()
|> IO.puts()

contents
|> Day01.solve_part2()
|> IO.puts()

File: day01/main.c
#include <aoc/aoc.h>
#include <aoc/mem.h>
#include <stdio.h>

static i32 solve_part1(const char *input) {
  i32 solution = 0;
  while (*input)
    solution += *input++ == '(' ? 1 : -1;
  return solution;
}

static i32 solve_part2(const char *input) {
  for (i32 i = 0, floor = 0;; ++i) {
    floor += input[i] == '(' ? 1 : -1;
    if (floor == -1)
      return i + 1;
  }
  return 0; // should not reach
}

int main(void) {
  char *input = aoc_file_read_all2("day01/input.txt");
  printf("%d\n%d\n", solve_part1(input), solve_part2(input));
  aoc_free(input);
}
